design/hardware-classification-controller/support-for-new-parameters-hwcc.md

```diff
@@ -0,0 +1,243 @@
+
+    Investigation Details: Below is sample introspection data for Disk.
+
+     1. HDD - Identified using `rotational` parameter's value as true
+      and `hctl` value as "0:0:N:0".  
+     2. SSD - Identified using `rotational` parameter's value as false
+      and `hctl` value as "0:0:N:0" or "N:0:0:0".
+     3. NVMe - Identified with the combination of `model` name and
+      `rotational` parameter's value as false.
+     4. RAID - Identified with the combination of
+
+        a) `rotational` parameter's value as true.
+
+        b) If `hctl` have value as "N:0:0:0" or "0:N:0:0".
+
+     ```yaml
+     hardware:
+        disks:
+          - name: "/dev/sda"
+            model: "KPM5XVUG3T84"
+            rotational: "true"
+            hctl: "0:0:N:0"
+     ```
+3. NIC
+
+     Adding new parameter `vendorName` under NIC. User can provide NIC
+      requirement in the form of `vendor`.
+
+     Investigation details: Below is sample introspection data for NICs.
+      `model` field under `nics` contains combination of Vendor ID and
+       Product ID.
+
+     ```yaml
+     hardware:
+       nics:
+       - ip: ""
+         mac: 24:6e:96:aa:bb:cc
+         model: 0x8086 0x1234
+         name: eth3
+         pxe: false
+     ```
+4. Firmware
+
+    Firmware is new parameter to be introduced in HardwareClassification
+     which contains BIOS information. User can specify BIOS requirements
+     by providing `vendor` and `version` details.
+
+    Investigation details: Below is sample introspection data for firmware.
+     `BIOS` field under firmware have two fields, `vendor` and `version`.
+
+    ```yaml
+    hardware:
+        firmware:
+             bios:
+               date: "10/17/2018"
+               vendor: "Dell Inc."
+               version: "1.5.6"
+    ```
+5. SystemVendor 
+
+    SystemVendor is new parameter to be introduced which helps user to 
+    classify hosts of certain manufacturer and product. User needs to 
+    provide `manufacturer` and `productName` in yaml.
+     
+    Investigation details: Below is sample introspection data for 
+    systemVendor. `systemVendor` field contains `manufacturer` and
+     `productName`. 
+   
+    ```yaml
+    hardware:
+        systemVendor:
+          manufacturer: Dell Inc.
+          productName: PowerEdge R640 (SKU=NotProvided;ModelName=PowerEdge R640)
+    ```
+
+### Implementation Details/Notes/Constraints
+ 
+Link for Existing HWCC Specs
+[Existing YAML](https://github.com/metal3-io/hardware-classification-controller/blob/master/config/samples/metal3.io_v1alpha1_hardwareclassification.yaml)
+ 
+* Below is sample yaml for additional parameters in HardwareClassification.
+ 
+   ```yaml
+   
+   apiVersion: metal3.io.sigs.k8s.io/v1alpha1
+   kind: HardwareClassification
+   metadata:
+     name: profile1
+   spec:
+     hardwareCharacteristics:
+       CPU:
+          arch: x86_64
+          minimumCount: 4
+          minimumSpeed: 4.3
+       Disk:
```

Comment:
Changing `disk` from a single field to a list like this is not backwards compatible, and that will increase the complexity of the implementation significantly because we will have to provide conversion a webhook and publish 2 versions of the CRD. I suggest instead that we add a new field to describe the combinations of disks. Maybe with a name like `diskPools`?

Is this example showing "and" or "or" rules? Does the host have to have 2 SSD and 1 RAID volume, or will a host with either of those match?

Reply:
yeah. I gave a thought on this. When it comes to multiple disks, current proposed solution may not work fully. 
When we add diskPools for all disk types, problem is how can we satisfy the minimum and maximum requirement for all the expected characteristics. Need to think more on this.

Reply:
Perhaps I misunderstood the intent of listing multiple items here under `disk`, and we just need a better name for the new field?

Reply:
This comment still needs to be addressed, I think? We cannot change `disk` to a list, so if we need a list we need a new field name and we need to describe which field takes precedence when both are present.

Reply:
Since the disk and NIC sections are removed above, you could remove them from this example, too.